# 122 statistics repository 구현 및 Dao 작성

WooVictory opened the pull request:
## 개요
- StatisticsRepository 구현 및 StatisticsDao 작성
## 작업사항
- StatisticsDao를 생성하고 쿼리를 작성
- StatisticsRepositoryImpl을 통해서 StatisticsRepository 인터페이스를 구현
- resolved #122

WooVictory committed:
issue-122-statistics repository 구현을 위한 StatisticsDao 작성

WooVictory committed:
issue-122-statistics repository 구현을 위한 StatisticsDao 작성

WooVictory committed:
issue-122-StatisticsRepositoryImpl 클래스 작성

WooVictory committed:
issue-122-StatisticsRepositoryImpl 클래스 작성

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/local/room/dao/StatisticsDao.java:
```diff
@@ -0,0 +1,21 @@
+import teamh.boostcamp.myapplication.data.model.CountedTag;
+import teamh.boostcamp.myapplication.data.model.EmotionHistory;
+
+public interface StatisticsDao {
```
room의 DAO로 작동하려면 `@Dao`가 있어야 합니다

WooVictory replied:
@Dao 추가했습니다.

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/local/room/dao/StatisticsDao.java:
```diff
@@ -0,0 +1,21 @@
+    @NonNull
+    @Query("SELECT diary.timeStamp, diary.selectedEmotion, diary.selectedEmotion " +
+            "FROM diary ORDER By timeStamp ASC LIMIT 14")
+    Single<List<EmotionHistory>> loadRecentEmotionHistoryList();
```
일단 쿼리 자체도 잘못되었구요 저렇게 값을 가져와서 자동으로 EmotionHistory로 mapping되지도 않습니다. EmotionHistory는 Room의 entity가 아닙니다..

room 사용법을 좀 더 공부해보세요.

그리고 이건 DiaryDao로 가는게 맞겠네요.

WooVictory replied:
DiaryDao로 수정했고 쿼리에 대한 지식이 부족하여 아는 한에서 수정했습니다. 그리고 Mapper를 어떻게 작성해야 할지 혼란이 와서 EmotionHistory관련 Mapper는 여쭤보고 작성하려고 했습니다.

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/local/room/dao/StatisticsDao.java:
```diff
@@ -0,0 +1,21 @@
+
+    @NonNull
+    @Query("SELECT diary.tags FROM diary ORDER By timeStamp ASC LIMIT 14")
+    Single<List<CountedTag>> loadRecentCountedTagList();
```
위와 마찬가지 입니다.

WooVictory replied:
쿼리와 return 타입을 수정했습니다. Mapper도 생성했습니다.

WooVictory committed:
issue-118-EmotionHistory 데이터 클래스 수정

WooVictory committed:
issue-118-EmotionHistory 데이터 클래스 수정

WooVictory committed:
issue-122-StatisticsDao에서 @Dao 어노테이션 추가 및 함수 DiaryDao로 이동

WooVictory committed:
issue-122-StatisticsDao에서 @Dao 어노테이션 추가 및 함수 DiaryDao로 이동

WooVictory committed:
issue-122-CountedTagMapper 생성
- 반환되는 tag를 CountedTagMapeer를 통해 Mapping 하도록 수정했습니다.
- StatisticsRepositoryImpl 에서 오버라이드 함수에서 map을 사용했습니다.
- StatisticsDao의 함수 반환값 수정

WooVictory committed:
issue-122-CountedTagMapper 생성
- 반환되는 tag를 CountedTagMapeer를 통해 Mapping 하도록 수정했습니다.
- StatisticsRepositoryImpl 에서 오버라이드 함수에서 map을 사용했습니다.
- StatisticsDao의 함수 반환값 수정

WooVictory committed:
issue-122-CountedTagMapper 생성
- StatisticsRepositoryImpl 수정
- # develop을 pull 받아서 이전 commit 이력들이 존재하는 commit

WooVictory committed:
issue-122-StatisticsRepository 수정

WooVictory committed:
issue-122-StatisticsRepository 수정

WooVictory committed:
issue-122-사용하지 않는 import 제거

WooVictory committed:
issue-122-사용하지 않는 import 제거

WooVictory committed:
issue-122-머지로 인해 발생한 충돌 해결

WooVictory committed:
issue-122-머지로 인해 발생한 충돌 해결

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/model/EmotionHistory.java:
```diff
@@ -4,22 +4,27 @@
-                          @NonNull Emotion selectedEmotion,
-                          @NonNull Emotion analyzedEmotion) {
+                          @NonNull Emotion emotion,
+                          int emotionType) {
```
값이 정해져 있는 것이므로 enum 타입을 하나 추가해 주세요

WooVictory replied:
enum 타입을 추가해서 수정했습니다.
https://github.com/boostcampth/boostcamp3_H/pull/124/commits/11786a9d80499ad2c4dc83b5bd5a9eaf51acc217

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/local/room/dao/DiaryDao.java:
```diff
@@ -19,5 +23,12 @@
 
     @Insert(onConflict = OnConflictStrategy.REPLACE)
-    void insert(@NonNull DiaryEntity...diaryEntities);
+    Completable insertDiary(LegacyDiary... diaries);
```
이게 갑자기 왜 추가 됬나요

WooVictory replied:
충돌 해결하는 과정에서 확인하지 못했습니다. 제거했습니다.

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/repository/CountedTagMapper.java:
```diff
@@ -0,0 +1,23 @@
+        for (int i = 0; i < size; i++) {
+            String tag = tagList.get(i);
+
+            countedTagList.add(new CountedTag(tag, i));
```
다시 한번 확인해 주세요

zeropol2 replied:
i가 무슨 의미가 있어서 CountedTag에 넣어줄까요? 저기는 가져온 리스트에서 해당 tag가 몇번이나 나왔는지 넣어줘야 하는거 아닐까요?

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/repository/StatisticsRepository.java:
```diff
@@ -5,20 +5,15 @@
 
-    @NonNull
-    Single<List<EmotionHistory>> loadRecentEmotionHistoryList();
-
```
이건 왜 지우셨어요?

WooVictory replied:
인터페이스를 구현함에 있어서 Mapping을 할 때 어떻게 해야 할 지 몰라서 일단 빼놨습니다..

zeropol2 replied:
다시 살리세요.. 구현을 해야 할 스펙이라고 정의해놨는데, 구현을 못하겠다고 빼시면 안됩니다..

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/local/room/dao/DiaryDao.java:
```diff
@@ -19,5 +23,12 @@
+
+    @NonNull
+    @Query("SELECT * FROM diary ORDER By timeStamp ASC LIMIT 14")
+    Single<List<DiaryEntity>> loadRecentEmotionHistoryList();
```
이름을 다시 생각해 주세요 
limit값은 파리미터로 빼주시구요

WooVictory replied:
limit 값은 파라미터로 뺐습니다.

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/local/room/dao/DiaryDao.java:
```diff
@@ -20,10 +23,12 @@
-    @Query("SELECT diary.timeStamp, diary.selectedEmotion, diary.selectedEmotion " +
-            "FROM diary ORDER By timeStamp ASC LIMIT 14")
-    Single<List<EmotionHistory>> loadRecentEmotionHistoryList();
+    @Query("SELECT * FROM diary ORDER By timeStamp ASC LIMIT 14")
```
DiaryEntity에 timeStamp라는 필드가 있나요?

zeropol2 replied:
Diary 테이블 명도 확인하시구요

WooVictory replied:
변경된 테이블명 확인하고 수정했습니다.

WooVictory committed:
issue-122- inserDiary 삭제

WooVictory committed:
issue-122
- DiaryDao 수정
- StatisticsDao 수정
- EmotionType enum 추가
- EmotionHistory에 enum 필드 추가

WooVictory committed:
issue-122
- AppDatabase에 StatisticsDao 추가
- StatisticsRepository 함수 추가

WooVictory committed:
issue-122- remote로부터 pull로 인한 충돌 해결

# Conflicts:
#	app/src/main/java/teamh/boostcamp/myapplication/data/local/room/dao/DiaryDao.java
#	app/src/main/java/teamh/boostcamp/myapplication/data/local/room/dao/StatisticsDao.java
#	app/src/main/java/teamh/boostcamp/myapplication/data/model/EmotionHistory.java
#	app/src/main/java/teamh/boostcamp/myapplication/data/repository/StatisticsRepositoryImpl.java

WooVictory committed:
issue-122-CountedTagMapper 충돌 해결

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/local/room/dao/StatisticsDao.java:
```diff
@@ -0,0 +1,19 @@
+
+    @NonNull
+    @Query("SELECT diaries.tags FROM diaries ORDER By recordDate ASC LIMIT :term")
+    Single<List<String>> loadRecentCountedTagList(int term);
```
이것도 diaryDao로 가야 되고, 이 StatisticsDao는 제거 해주세요

zeropol2 replied:
이거 제대로 안가져와질것 같은데요, 그냥 diarylist를 가져온다음에 거기서 tagList를 가져오는 방식으로 하시면 될것 같으니 이건 그냥 제거해주시고 DiaryDao에 있는 loadDiaryList를 쓰면 될것 같습니다.

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/local/room/dao/StatisticsDao.java:
```diff
@@ -0,0 +1,19 @@
+public interface StatisticsDao {
+
+    @NonNull
+    @Query("SELECT diaries.tags FROM diaries ORDER By recordDate ASC LIMIT :term")
```
최근 걸 가져오시려면 정렬이 날짜 내림차순이어야 하지 않을까요?

WooVictory committed:
issue-122-StatisticsDao 삭제 및 DiaryDao에 함수 추가

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/model/EmotionType.java:
```diff
@@ -0,0 +1,17 @@
+
+    selectedEmotion(0), analyzedEmotion(1);
+
+    private final int emotionType;
```
굳이 value가 필요 없을 것 같은데, 어떤 이유로 만드신 것인지요?
value없이 그냥 enum값만 있으면 될 것 같아서요

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/repository/CountedTagMapper.java:
```diff
@@ -0,0 +1,23 @@
+
+            countedTagList.add(new CountedTag(tag, i));
+        }
+        return countedTagList;
```
이 로직이 원하시는 로직이 맞는건지 다시 한번 확인해보시기 바랍니다.

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/repository/StatisticsRepositoryImpl.java:
```diff
@@ -0,0 +1,41 @@
+    @NonNull
+    @Override
+    public Single<List<CountedTag>> loadRecentCountedTagList() {
+        return appDatabase.statisticsDao().loadRecentCountedTagList(14)
```
14를 별도의 상수로 빼시죠

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/local/room/dao/DiaryDao.java:
```diff
@@ -19,5 +20,10 @@
+
+    @NonNull
+    @Query("SELECT * FROM diaries ORDER By diaries.recordDate ASC LIMIT :term")
+    Single<List<DiaryEntity>> loadRecentEmotionHistoryList(int term);
```
여기서는 EmotionHistory를 가져오는게 아니고 그냥 DiaryEntity List를 가져오는 겁니다. repository단에서 DiaryEntity를 EmotionHistory로 가공하는 거구요..
이것도 그냥 loadDiaryList로 쓰고 이건 지워도 될것 같아요

WooVictory committed:
issue-122-enumType 수정

WooVictory committed:
issue-122-DiaryDao에서 함수들 제거

WooVictory committed:
issue-122-Repository에서 감정 리스트 처리

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/repository/StatisticsRepositoryImpl.java:
```diff
@@ -0,0 +1,70 @@
+    @NonNull
+    final AppDatabase appDatabase;
+    private List<DiaryEntity> diaryEntityList;
+    private List<EmotionHistory> totalEmotionHistoryList;
```
이 두개 리스트는 field로 들고 있으면 안됩니다.

zeropol2 replied:
메소드 안으로 옮겨주세요

WooVictory replied:
지역변수로 변경했습니다.

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/repository/StatisticsRepositoryImpl.java:
```diff
@@ -0,0 +1,70 @@
+    @Override
+    public Single<List<CountedTag>> loadRecentCountedTagList(
+            @NonNull Date lastItemSavedTime) {
+        return null;
```
이거 구현도 하셔야 합니다.

WooVictory replied:
구현했습니다.

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/repository/StatisticsRepositoryImpl.java:
```diff
@@ -0,0 +1,70 @@
+        totalEmotionHistoryList.addAll(selectedEmotionHistoryArrayList);
+        totalEmotionHistoryList.addAll(analyzedEmotionHistoryArrayList);
+
+        return (Single<List<EmotionHistory>>) totalEmotionHistoryList;
```
RxJava를 좀 찾아보셔야겠네요

WooVictory replied:
rxJava를 처음 접하다 보니 잘 와닿지 않아서 찾아보면서 수정하도록 하겠습니다..!

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/repository/StatisticsMapper.java:
```diff
@@ -0,0 +1,28 @@
+        }
+        return emotionHistories;
+    }
+}
```
사용처가 없습니다. 지우세요

WooVictory replied:
삭제했습니다.

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/repository/CountedTagMapper.java:
```diff
@@ -0,0 +1,23 @@
+        }
+        return countedTagList;
+    }
+}
```
이것도 그냥 지우시고 필요하시면 만드세요.

WooVictory replied:
삭제했습니다.

WooVictory committed:
issue-122-Repository 구현

WooVictory committed:
issue-122- Mapper 삭제
